src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -1226,19 +1005,73 @@ public int hashCode() {
         }
     }
 
-    private void logNodeMessage(FlowNode node) {
-        WorkflowConsoleLogger wfLogger = new WorkflowConsoleLogger(getListener());
-        String prefix = getBranchName(node);
-        if (prefix != null) {
-            wfLogger.log(String.format("[%s] %s", prefix, node.getDisplayFunctionName()));
-        } else {
-            wfLogger.log(node.getDisplayFunctionName());
+    /**
+     * Prints nodes as they appear (including block start and end nodes).
+     */
+    private final class NodePrintListener implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            NewNodeConsoleNote.print(node, getListener());
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override public AnnotatedLargeText getLogText() {
+        return LogStorage.of(asFlowExecutionOwner()).overallLog(this, !isLogUpdated());
+    }
+
+    // TODO log-related overrides pending JEP-207:
+
+    @Override public InputStream getLogInputStream() throws IOException {
+        // Inefficient but probably rarely used anyway.
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        getLogText().writeRawLogTo(0, baos);
+        return new ByteArrayInputStream(baos.toByteArray());
+    }
+
+    @Override public Reader getLogReader() throws IOException {
+        return getLogText().readAll();
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override public String getLog() throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        getLogText().writeRawLogTo(0, baos);
+        return baos.toString("UTF-8");
+    }
+
+    @Override public List<String> getLog(int maxLines) throws IOException {
+        int lineCount = 0;
+        List<String> logLines = new LinkedList<>();
```

Comment:
🐜 [Why you should never use a LinkedList](https://kjellkod.wordpress.com/2012/02/25/why-you-should-never-ever-ever-use-linked-list-in-your-code-again/).  TL;DR: Locality-of-reference matters, allocate a darned `ArrayList`.

Reply:
This is just copied from core.

Reply:
Note that the core implementation of _this_ method has since been rewritten. As I pointed out in https://github.com/jenkinsci/jenkins/pull/2607#issuecomment-425475683 that is irrelevant, since this method is not used in production workflows. Anyway the rewrite is impossible to apply here since it relies on `RandomAccessFile`.

Reply:
Copy and pasted code?  Still... 

I'm going to keep pointing out that LinkedList is a bad idea until people stop (ab)using it.

Reply:
This method is not used in production. The performance does not matter.